### ISSUE TITLE
Retrofitted PoV Weights For Dispatchables

### ISSUE THREAD

**Issue by ggwpez** [paritytech/substrate#13501]
The PoV benchmarking aims to estimate the worst-case scenario, which is unlikely to occur on average.  
Too large over-estimations hurts block utilization. There is currently no way to retroactively measure the actual used proof size.  
Having such a retroactive measurement in place would enable a multi iteration approach; alternating between dispatch and assessment phase to reach maximum block utilization.

It would therefore be nice to have a way to measure the weight of a batch of dispatchables after their execution. For example in:
- Scheduler Task dispatch
- XCM dispatch
- Transaction dispatch
- MessageQueue processing

Basically in every situation where we need to fit a batch of Dispatchables into a weight limit and assume that the pre- and post-dispatch PoV weight differ greatly.

**Comment by kianenigma** [comment]
@paritytech/sdk-node would be good to get some input here how the proof recording is currently working in the client side, and what else can be done to assist this. 

For example, how expensive is it to know the accurate proof as we are building the block? Can the client measure and report back the actual post-dispatch (`proof_size`) weight?

**Comment by xlc** [comment]
we can’t realistically enable proof size benchmarks as current without big regression on the block size. this is a major blocking issue.

**Comment by ggwpez** [comment]
We could start with the compacted proof, that should already bring advantages in cases that keys are accessed repetitively.  
Otherwise we need some stream compression? I remember reading an issue/discussion about it.

**Comment by bkchr** [comment]
I think the better solution and much easier is to come up with some kind of factor between the storage proof and the compressed storage proof. @cheme was always arguing that the compression isn't bringing that much as we have binary data that can not be compressed that much anymore. So, maybe we should collect all the factors of storage proof vs compressed in the benchmarking and then taking the worst factor to calculate the compressed storage proof size based on the storage proof size returned by the host function.

**Comment by xlc** [comment]
Some of the data can be compressed well such as `Balance` and some don't (`AccountId`) so the compress ratio could vary by the usage.

But I think the main contributor of big size is difference between `MaxEncodedLen` and actual average item size

**Comment by cheme** [comment]
> > For example, how expensive is it to know the accurate proof as we are building the block? Can the client measure and report back the actual post-dispatch (`proof_size`) weight?
> 
> Relative "cheap":
> 
> https://github.com/paritytech/substrate/blob/5b253f8757de20cd95f274eb35531a5b0a683987/primitives/trie/src/recorder.rs#L155-L157
> 
> However, we currently don't take the compressed size into account.

This is also not estimating the compacted proof (I did try to have primitive for compacted proof, it means recording parent-child relation too in order to reduce estimated size, it was not too simple or cheap so I drop the branch).

For compression, it does not compress much on all that is trie nodes (morstly crypto hash), but the values in the proof can be compressed (not all value of course, but it depends on the chain here). But there might be something to do by compressing the value only for estimation (at least a proof format could be design with two files: one with trie material and one with indexed valuers, then we only compress the second one).

**Comment by bkchr** [comment]
I would propose that we check the current storage proof size after applying an extrinsic. Then we would return the unused (worst_case - actual) proof size as we return the ref time. This should be implemented in the following way: 

1. We introduce a custom host function into Cumulus:
```
/// Returns the current uncompressed storage proof size.
///
///  The storage proof size is composed by the size of all trie nodes/values read from the state.
fn storage_proof_size() -> u32; 
``` 

In Substrate this will mean that we will introduce a new [`Extension`](https://docs.rs/sp-externalities/latest/sp_externalities/trait.Extension.html) that will hold a reference to the `StorageProof` to call `estimate_encoded_size()`. When this extension isn't registered, we the host function should return 0. 

While the host function isn't used by Polkadot itself, we should still create some RFC for adding it.

2. There are four different "stages" in which this host function will be called.

2.1. Block building

When we are building a block, we already collecting the storage proof. So, we will just need to register the extension and it should work.

2.2 Block import

Currently we don't collect a storage proof while importing a block, as we don't need it. However, we would need to start doing this to have `storage_proof_size` return the correct values while importing a block. This would be the simplest solution to the problem and should also be very easy to implement for light clients (They would just need to return the size of the storage proof they have fetched so far, when they actually need to execute a block).

Another more complex solution would be to alter FRAME to have some new hook `on_final_finalize` (name is a joke, naming is hard :P). The idea of this hook would be that you are only allowed to do a really small subset of actions that require extra caution, but you could return a vec of custom `Calls` that can be attached to the block. This would run after `on_finalize` and before we calculate the storage root etc. These custom `Calls` would then be added as unsigned `Extrinsic` to the block (would require to change the `BlockBuilder` runtime api to also have these extrinsic being returned). Before executing the block in `execute_block` we would then remove these extrinsics from the block and setup certain things based on them. In our use case here we would store the return values of `storage_proof_size` as a vector in a call. When we would then call `execute_block` it would take the vec of values and setup of `storage_proof_size` to not call into the host and stay in the runtime and return always the next value in the vector. (The way of adding a custom calls/extrinsics by the runtime to the block is also something I would always had imagined for: https://github.com/paritytech/cumulus/issues/320)

2.3 Block validation

Depending how we implement the solution for block import, block validation would do it similar. If we enable storage proof recording while block import, it would require to add some new field to [`ParachainsBlockData`](https://github.com/paritytech/cumulus/blob/8ee8dc11e8c123c4d94ca9d10451f267d7ec3f33/primitives/core/src/lib.rs#L157-L164). This field would be a `Vec<u32>` and each entry corresponds to the return value of `estimate_encoded_size` while building the block. We would then use this vec to "replay" the same return values as we are running the validation. This will be done by overwriting the `storage_proof_size` host function as we do it for the storage host functions etc.

If we choose the other solution as outlined above, block validation will not require any kind of special handling and will just work™

2.4 Offchain calls

For any kind of offchain call that maybe is hitting `storage_proof_size` it should return `0`.

We now only need to discuss on which block import solution to implement.

I would propose @skunert as the one implementing the changes to make the host function work. The actual FRAME integration should then be relative easy.

**Comment by bkchr** [comment]
> I thought the compacted proof is just de-duplicating node entries from the proof? Or is that already happening?

We don't have duplicated nodes by default. This is by design. Compaction removes an intermediate node when we have all the childs available to recreate it.


> Is there an easy way to keep host functions extensible?

I'm not really convinced that we need this. Especially if we would add a new enum variant, we should also bump the host function version. 

Stuff like compaction could probably be implemented using some kind of iterative algorithm. However, we would then still need to compress the compact result which can not be iterative, as we could have removed a node in between using compaction.

> I assume this host-function is transactional aware

The storage recorder is transactional, but this isn't working as you expect it. We can only discard a layer if the extrinsic isn't part of the block. The moment the transaction is part of the block, we will need to record all the accessed nodes and can not throw away anything. Otherwise the validation will not work.

> could just called within something like `Utility::batch` to clawback the unused PoV of each call when looping over them?

It should work, but I don't see the reason? You can just do it after `batch` is finished? As before you run `batch` you need to take the static weight any way to decide if the extrinsic/calls fits into the block. In this moment you already remove all the static weight and can refund later. No need to return in between.

> The MQ pallet would also profit from this, so that it can reclaim unused PoV and process more messages.

That makes sense. I think in general you didn't understand what I meant with these "stages". These are seen from the node side, the runtime doesn't care and can call the host function as it likes.

**Comment by ggwpez** [comment]
>It should work, but I don't see the reason? You can just do it after batch is finished?

Ah yea, that was a stupid example :see_no_evil: 

>>The MQ pallet would also profit from this, so that it can reclaim unused PoV and process more messages.

>That makes sense. I think in general you didn't understand what I meant with these "stages". These are seen from the node side, the runtime doesn't care and can call the host function as it likes.

Okay. I was only concerned that it would not update the PoV estimation immediately after a storage access in the runtime, but it seems to be the case.

**Comment by bkchr** [comment]
The `PovRecorder` records as you access the storage and the POV size estimation is also always updated.

**Comment by gavofyork** [comment]
So it's not just extrinics, but also `on_initialize`/`on_idle` code across pallets, the scheduler and the message-queue, as well as any ecosystem pallets which also do dynamic stuff in `on_initialize`/`on_idle`. So we should be careful not to over-specialise for the blockbuilder/extrinsic use-case.

Jacob mentioned correctly that we don't need to correct-down the weight tracker on every operation (extrinsic/task/message), but only when we believe we have reached some limit. At this point and only here shoudl we correct-down to the actual amount of weight used and recheck against the limit properly.

In the case of scheduler to which a chain might allocate 80% of block space, the first five tasks might use, say, 75%, with the sixth task estimated at 10%. It would only be at this point that the weight is corrected-down to, say, 45% and the sixth task executed.

**Comment by gavofyork** [comment]
@bkchr I presume "Block import" is what happens in the collator node, whereas "Block validation" is what happens in the validator node?

Seems sensible to go the `Vec<u8>` way - but why would it not be done at the point of block builder and distributed alongside the block?

**Comment by bkchr** [comment]
> @bkchr I presume "Block import" is what happens in the collator node, whereas "Block validation" is what happens in the validator node?

Yeah, sorry! I should have been more clear. By block import I mean when any node imports this block to its database. By block validation I meant the relay chain running `validate_block` for a given PoV.

> Seems sensible to go the `Vec<u8>` way

What you mean by the `Vec<u8>` way?

> but why would it not be done at the point of block builder and distributed alongside the block?

We don't support distributing any data alongside the block, at least not currently. In the database we store the parts of the block separately, aka the `Header` and the `Extrinsics`. Thus, my idea was that we add the information to "reconstruct" `storage_proof_size` into a transaction. However, this extrinsic needs to be generated inside the runtime to ensure that the `extrinsics_root` is correct and it also needs to be generated at a point at which no one is accessing this functionality anymore.

**Comment by bkchr** [comment]
I briefly talked with @gavofyork.

We came to the following conclusion/idea:

- We record all the accesses to the `storage_proof_size` host function on the host side.
- After we applied all the extrinsics there will be a new "inherent phase". So, we call into the runtime again, passing it data to construct inherent extrinsics. These inherents are then pushed at the end of the block. These inherents should not do any kind of computation, only contain our data.
- The data is then shared alongside the block. When importing the block or validating the PoV we will then have the data available and can use it to reconstruct the calls to the host function.
- It will not be allowed to call `storage_proof_size` as part of `on_finalize` as this would not be recorded.
- We will have some chicken/egg problem as the data we collect to construct this inherent actually also contributes to the PoV size. But it shouldn't be such a big problem and we can just reserve some space for this data in our calculations.

**Comment by gavofyork** [comment]
@bkchr we also briefly discussed the possibility of (re)evaluating the storage proof size during execution on later validation/execution by marking the parts of the proof which get used as they get entered into a cache. This would obviate the need for this extra inherent phase, the new requirements of `finalize_block` and the "data extrinsic".

**Comment by bkchr** [comment]
The idea was that in `validate_block` we could use the [`TrieCache`](https://github.com/paritytech/cumulus/issues/2432) to find out what the proof size is. As the cache represents all the de-duplicated data that we have read, requesting its size should work as return value of `storage_proof_size`.

The only question would be on what to do with `execute_block` on full nodes. The best and most easiest solution would probably to enable storage proof recording to be able to answer the `storage_proof_size` requests. I think we can go with this.

**Comment by kianenigma** [comment]
Once we have this, we can do some kind of a "auto-refund" scheme for all dispatchable's `proof_size` dimension. This won't be make sense for `ref_time`, but if we know the actual proof size post dispatch, we can automatically refund it. 

This will help increase the block utilization, and slightly reduce fees for end users.
